scrapy/downloadermiddlewares/cookies.py

```diff
@@ -69,26 +69,60 @@ def _debug_set_cookie(self, response, spider):
                 msg = "Received cookies from: {}\n{}".format(response, cookies)
                 logger.debug(msg, extra={'spider': spider})
 
-    def _format_cookie(self, cookie):
-        # build cookie string
-        cookie_str = '%s=%s' % (cookie['name'], cookie['value'])
-
-        if cookie.get('path', None):
-            cookie_str += '; Path=%s' % cookie['path']
-        if cookie.get('domain', None):
-            cookie_str += '; Domain=%s' % cookie['domain']
-
+    def _format_cookie(self, cookie, request):
+        """
+        Given a dict consisting of cookie components, return its string representation.
+        Decode from bytes if necessary.
+        """
+        decoded = {}
+        for key in ('name', 'value', 'path', 'domain'):
+            if not cookie.get(key):
+                if key in ('name', 'value'):
+                    msg = 'Invalid cookie found in request {}: {} (\'{}\' is missing)'.format(request, cookie, key)
+                    logger.warning(msg)
+                    return
+                continue
+            if isinstance(cookie[key], six.text_type):
+                decoded[key] = cookie[key]
+            else:
+                try:
+                    decoded[key] = cookie[key].decode('utf8')
+                except UnicodeDecodeError:
+                    logger.warning('Non UTF-8 encoded cookie found in request %s: %s', request, cookie)
+                    decoded[key] = cookie[key].decode('latin1', errors='replace')
+
+        cookie_str = u'{}={}'.format(decoded.pop('name'), decoded.pop('value'))
+        for key, value in six.iteritems(decoded):  # path, domain
+            cookie_str += u'; {}={}'.format(key.capitalize(), value)
         return cookie_str
 
     def _get_request_cookies(self, jar, request):
-        if isinstance(request.cookies, dict):
-            cookie_list = [{'name': k, 'value': v} for k, v in \
-                    six.iteritems(request.cookies)]
-        else:
-            cookie_list = request.cookies
-
-        cookies = [self._format_cookie(x) for x in cookie_list]
-        headers = {'Set-Cookie': cookies}
-        response = Response(request.url, headers=headers)
-
-        return jar.make_cookies(response, request)
+        """
+        Extract cookies from a Request. Values from the `Request.cookies` attribute
+        take precedence over values from the `Cookie` request header.
+        """
+        def get_cookies_from_header(jar, request):
+            cookie_header = request.headers.get('Cookie') or b''
+            cookie_list_bytes = re.split(b';\s*', cookie_header)
```

Comment:
`_get_request_cookies` is on every request path, considering that we should precompile regexes and exit earlier if `Cookie` header is not set.

Reply:
Makes a lot of sense, thanks.

Reply:
@dangra Done, thanks for the suggestion!